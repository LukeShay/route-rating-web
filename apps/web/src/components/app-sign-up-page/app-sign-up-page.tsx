import { Component, Host, h } from '@stencil/core'

import { HTMLEvent } from '@routerating/components'
import authStore from '../../stores/auth'
import { createBasicUser } from '@routerating/requests'
import toastStore from '../../stores/toast'

interface Refs {
  firstName?: HTMLRrInputElement
  lastName?: HTMLRrInputElement
  email?: HTMLRrInputElement
  phoneNumber?: HTMLRrInputElement
  city?: HTMLRrInputElement
  state?: HTMLRrInputElement
  password?: HTMLRrInputElement
  repeatPassword?: HTMLRrInputElement
}

@Component({
  tag: 'app-sign-up-page',
  styleUrl: 'app-sign-up-page.scss',
  shadow: true,
})
export class AppSignUpPage {
  private refs: Refs = {}

  render() {
    return (
      <Host>
        <app-form>
          <h1 class="centered">Sign Up</h1>
          <form onSubmit={this.handleSubmit}>
            <rr-input
              label="First name"
              type="text"
              required
              ref={(element?: HTMLRrInputElement): void => {
                this.refs.firstName = element
              }}
            />
            <rr-input
              label="Last name"
              type="text"
              required
              ref={(element?: HTMLRrInputElement): void => {
                this.refs.lastName = element
              }}
            />
            <rr-input
              label="Email"
              type="text"
              required
              ref={(element?: HTMLRrInputElement): void => {
                this.refs.email = element
              }}
            />
            <rr-input
              label="City"
              type="text"
              required
              ref={(element?: HTMLRrInputElement): void => {
                this.refs.city = element
              }}
            />
            <rr-input
              label="State"
              type="text"
              required
              ref={(element?: HTMLRrInputElement): void => {
                this.refs.state = element
              }}
            />
            <rr-input
              label="Phone number"
              type="text"
              required
              ref={(element?: HTMLRrInputElement): void => {
                this.refs.phoneNumber = element
              }}
            />
            <rr-input
              label="Password"
              type="password"
              required
              ref={(element?: HTMLRrInputElement): void => {
                this.refs.password = element
              }}
            />
            <rr-input
              label="Repeat password"
              type="password"
              required
              ref={(element?: HTMLRrInputElement): void => {
                this.refs.repeatPassword = element
              }}
            />
            <div class="centered">
              <button type="submit" class="purple-button">
                Sign Up
              </button>
            </div>
          </form>
        </app-form>
      </Host>
    )
  }

  private handleSubmit = async (event: HTMLEvent): Promise<void> => {
    event.preventDefault()

    if (
      (await this.refs.password?.getValue()) !==
      (await this.refs.repeatPassword?.getValue())
    ) {
      toastStore.message = 'Passwords must match.'
      return
    }

    const response = await createBasicUser('http://localhost:3000/dev', {
      firstName: (await this.refs.firstName?.getValue()) ?? '',
      lastName: (await this.refs.lastName?.getValue()) ?? '',
      email: (await this.refs.email?.getValue()) ?? '',
      phoneNumber: (await this.refs.phoneNumber?.getValue()) ?? '',
      city: (await this.refs.state?.getValue()) ?? '',
      state: (await this.refs.state?.getValue()) ?? '',
      password: (await this.refs.password?.getValue()) ?? '',
    })

    if (response.ok) {
      authStore.authToken = response.headers.get('Authorization')
      authStore.refreshToken = response.headers.get('Refresh')
      authStore.signedIn = response.headers.get('Authorization') !== null

      toastStore.message = 'You have been signed in.'
    } else {
      toastStore.message = 'Unable to create an account.'
    }
  }
}
