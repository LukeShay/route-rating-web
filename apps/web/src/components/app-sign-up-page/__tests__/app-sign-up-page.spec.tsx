import { AppSignUpPage } from '../app-sign-up-page'
import { newSpecPage } from '@stencil/core/testing'

describe('app-sign-up-page', () => {
  it('should have match snapshot', async () => {
    const page = await newSpecPage({
      components: [AppSignUpPage],
      html: `<app-sign-up-page></app-sign-up-page>`,
    })
    expect(page.root).toMatchSnapshot()
  })
})
