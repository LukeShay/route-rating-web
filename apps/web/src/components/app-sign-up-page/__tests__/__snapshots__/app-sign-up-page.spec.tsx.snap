// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app-sign-up-page should have match snapshot 1`] = `
<app-sign-up-page>
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
</app-sign-up-page>
`;
